src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const pipeline = new Pipeline(p.name);
```

Comment:
I wonder if we need to provide a `.clone` facility for all these objects. This "copy" of the pipeline is not correct, as it drops the `env` parameters